src/blackdoor/cqbe/node/server/RPCHandler.java

```diff
@@ -45,189 +43,206 @@
+								"WTF IS THISSSS??? I'm looking at a method type that I don't recognize!"
+										+ "WHERE is the validator? Is it on vacation? Cause it's not validating!");
+				}
+
+
 		} catch (JSONException j) {
+			// TODO not convinced this is needed anymore
 			DBP.printException(j);
 			DBP.printerrorln("Apparently the RPC validator is broken");
 			DBP.printerrorln("A JSON-RPC response is not being sent, better fix the validator");
 			return null;
-		} catch (AddressException a) {
-			DBP.printException(a);
-			responseObject = RPCBuilder.RPCResponseFactory(rpc.getInt("id"),
-					false, null,
-					RPCException.JSONRPCError.INVALID_ADDRESS_FORMAT,
-					a.getMessage());
 		} catch (RPCException e) {
-			if (errorData != null)
-				responseObject = RPCBuilder.RPCResponseFactory(
-						rpc.getInt("id"), false, null, e.getRPCError(),
-						errorData);
+			if (request == null && errorData != null)
+				response = new ErrorRpcResponse(e.getRPCError(), errorData);
+			else if (request != null && errorData == null)
+				response = new ErrorRpcResponse(request, e.getRPCError());
 			else
-				responseObject = RPCBuilder.RPCResponseFactory(
-						rpc.getInt("id"), false, null, e.getRPCError());
-		} catch (UnknownHostException e) {
-			responseObject = RPCBuilder.RPCResponseFactory(rpc.getInt("id"),
-					false, null,
-					RPCException.JSONRPCError.INVALID_ADDRESS_FORMAT);
+				response = new ErrorRpcResponse(e.getRPCError());
+		} catch (AddressException | UnknownHostException e) {
+			response =
+					new ErrorRpcResponse(request, RPCException.JSONRPCError.INVALID_ADDRESS_FORMAT);
+
 			DBP.printException(e);
+		} catch (IOException e) {
+			// TODO What do? shutdown had issues.....
 		}
-			// DBP.printdevln("in handle");
-			// DBP.printdevln("about to write response " + responseObject);
-
-		return responseObject;
-	}
 
-	/**
-	 * adds sender ip and port from rpc to this nodes's address table if
-	 * applicable. DOES NOT CHECK validity
-	 * 
-	 * @throws JSONException
-	 * @throws UnknownHostException
-	 */
-	private void addRequestSenderToAT() throws UnknownHostException{
-		JSONObject params = rpc.getJSONObject("params"); 
-		L3Address sender = new L3Address(InetAddress.getByName(params.getString("sourceIP")), params.getInt("sourcePort"));
-		if(!L3Address.isNonNodeAddress(sender))
-		{
-			L3Address result = Node.getAddressTable().add(sender);
-			if(!sender.equals(result))
-				DBP.printdemoln("Adding " + sender + " to address table from handler");
-		}
+		return response;
 	}
 
 	/**
-	 * Sends response to server, to be sent to the RPC builder
+	 * <p>
+	 * Checks for semantics, syntax and if the RPC is supported by this system.
+	 *
+	 * @param String
+	 * @return String detailing whether the JSONObject is valid or not.
 	 */
-	private void sendRPC() {
-
+	public boolean isValid(Rpc request) {
+		// TODO there are probably more things that make an RPC valid that arent
+		// being checked here...
+		if (request != null)
```

Comment:
this needs to be "if(request == null) return false;". problems will be had otherwise.